ext/mbstring/mbstring.c

```diff
@@ -5155,7 +5155,12 @@ static bool mb_fast_check_utf8_avx2(zend_string *str)
 			goto check_operand;
 		case 7:
 		case 8:
-			operand = _mm256_set_epi64x(0, 0, 0, *((int64_t*)p));
+			/* This was originally: operand = _mm256_set_epi64x(0, 0, 0, *((int64_t*)p));
+			 * However, it was found that caused test failures on 32-bit MS Windows
+			 * (Bad 7/8-byte UTF-8 strings would be wrongly passed through as 'valid')
+			 * It is not clear whether the root cause is a problem with MS Visual C++'s handling
+			 * of AVX2 intrinsics when building 32-bit binaries, or something else */
+			operand = _mm256_set_epi32(0, 0, 0, 0, 0, 0, ((int32_t*)p)[1], ((int32_t*)p)[0]);
```

Comment:
same instruction is used in https://github.com/php/php-src/blob/76386405291f41d9bd804f25ed7f55178073b98d/ext/hash/xxhash/xxhash.h#L3895

Reply:
Indeed.

I don't know if the AVX2-accelerated code you are showing from xxhash is used on 32-bit Windows hosts or not. If it is, that indicates that using `_mm256_set_epi64x` is not necessarily a problem in of itself. Whatever the case, a root-cause analysis cannot be done without access to a test machine, and I don't have one. However, the change in this PR *seems* to have fixed a problem.

If anyone else has more insight into the issue at hand, I would love to hear their comments.

Reply:
Should we use an `#if` and keep using `_mm256_set_epi64x` for everything but Windows x86 32-bit?

Reply:
seems to be https://stackoverflow.com/questions/37509129/potential-bug-in-visual-studio-c-compiler-or-in-intel-intrinsics-avx2-mm256-s and xxhash should be fixed as well

Reply:
@mvorisek Good find, that explains it 🙂

Reply:
@mvorisek Great find!!

As a suggestion, I think the first thing would be to add a test case which exposes the problem on 32-bit Windows. Then it will be easy to verify the fix.

Reply:
> Should we use an `#if` and keep using `_mm256_set_epi64x` for everything but Windows x86 32-bit?

I don't think so.

If this particular line was performance-critical, then I would say yes, but I think the performance loss here will be minute. (If the compiler is smart enough to generate optimal code on x86-64, we won't lose anything.) The effect will only be on 7 and 8 byte UTF-8 strings, and for those, I expect the overhead of entering/exiting a function from PHP code will dominate this tiny bit of overhead.

Having a lot of `#if`s in C code is a bad thing. I'm sure you know all the reasons why that is so.

Reply:
@iluuu1994 Please note, I would be happy to be proved wrong on the above statement.

Reply:
I wouldn't consider an `#if` an issue, but I'm fine either way :slightly_smiling_face: